third_party/move/move-vm/runtime/src/ops.rs

```diff
@@ -0,0 +1,256 @@
+
+use move_binary_format::file_format::{
+    Bytecode, CodeOffset, ConstantPoolIndex, FieldHandleIndex, FieldInstantiationIndex,
+    FunctionHandleIndex, FunctionInstantiationIndex, LocalIndex, SignatureIndex,
+    StructDefInstantiationIndex, StructDefinitionIndex, StructVariantHandleIndex,
+    StructVariantInstantiationIndex, VariantFieldHandleIndex, VariantFieldInstantiationIndex,
+};
+use move_core_types::{
+    function::ClosureMask,
+    int256::{I256, U256},
+};
+
+/// The VM's internal representation of instructions.
+///
+/// Currently, it is an exact mirror of the Move bytecode, but we plan to add more in the future.
+///
+/// This provides path for incremental performance optimizations, while making it less painful to
+/// maintain backward compatibility.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum Instruction {
+    Pop,
+    Ret,
+    BrTrue(CodeOffset),
+    BrFalse(CodeOffset),
+    Branch(CodeOffset),
+    LdU8(u8),
+    LdU64(u64),
+    LdU128(Box<u128>),
+    CastU8,
+    CastU64,
+    CastU128,
+    LdConst(ConstantPoolIndex),
+    LdTrue,
+    LdFalse,
+    CopyLoc(LocalIndex),
+    MoveLoc(LocalIndex),
+    StLoc(LocalIndex),
+    Call(FunctionHandleIndex),
+    CallGeneric(FunctionInstantiationIndex),
+    Pack(StructDefinitionIndex),
+    PackGeneric(StructDefInstantiationIndex),
+    PackVariant(StructVariantHandleIndex),
+    PackVariantGeneric(StructVariantInstantiationIndex),
+    Unpack(StructDefinitionIndex),
+    UnpackGeneric(StructDefInstantiationIndex),
+    UnpackVariant(StructVariantHandleIndex),
+    UnpackVariantGeneric(StructVariantInstantiationIndex),
+    TestVariant(StructVariantHandleIndex),
+    TestVariantGeneric(StructVariantInstantiationIndex),
+    ReadRef,
+    WriteRef,
+    FreezeRef,
+    MutBorrowLoc(LocalIndex),
+    ImmBorrowLoc(LocalIndex),
+    MutBorrowField(FieldHandleIndex),
+    MutBorrowVariantField(VariantFieldHandleIndex),
+    MutBorrowFieldGeneric(FieldInstantiationIndex),
+    MutBorrowVariantFieldGeneric(VariantFieldInstantiationIndex),
+    ImmBorrowField(FieldHandleIndex),
+    ImmBorrowVariantField(VariantFieldHandleIndex),
+    ImmBorrowFieldGeneric(FieldInstantiationIndex),
+    ImmBorrowVariantFieldGeneric(VariantFieldInstantiationIndex),
+    MutBorrowGlobal(StructDefinitionIndex),
+    MutBorrowGlobalGeneric(StructDefInstantiationIndex),
+    ImmBorrowGlobal(StructDefinitionIndex),
+    ImmBorrowGlobalGeneric(StructDefInstantiationIndex),
+    Add,
+    Sub,
+    Mul,
+    Mod,
+    Div,
+    BitOr,
+    BitAnd,
+    Xor,
+    Or,
+    And,
+    Not,
+    Eq,
+    Neq,
+    Lt,
+    Gt,
+    Le,
+    Ge,
+    Abort,
+    Nop,
+    Exists(StructDefinitionIndex),
+    ExistsGeneric(StructDefInstantiationIndex),
+    MoveFrom(StructDefinitionIndex),
+    MoveFromGeneric(StructDefInstantiationIndex),
+    MoveTo(StructDefinitionIndex),
+    MoveToGeneric(StructDefInstantiationIndex),
+    Shl,
+    Shr,
+    VecPack(SignatureIndex, u64),
```

Comment:
I think we can make this u64 be u16, bytecode verifier checks for bounds on this one (same for unpack, if you search for these instructions it is somewhere in bound checking or so) - might decrease instruction size further?